proto/prostgen/lib.rs

```diff
@@ -0,0 +1,194 @@
+mod extern_path;
+mod module;
+
+use extern_path::{ExternPath, ExternPathSetArg};
+use module::Module;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "Prostgen", about = "Generates protobuf code using prost")]
+pub struct ProstGen {
+    /// The paths to the proto files to be processed.
+    #[structopt(min_values = 1, parse(from_os_str))]
+    input: Vec<PathBuf>,
+
+    /// The directory where the output should be generated.  This will include a "lib.rs" as well as
+    /// all of the prost generated include!-able output.
+    #[structopt(long, short, parse(from_os_str))]
+    output: PathBuf,
+
+    /// The directories to be included in the invocation of protoc.
+    #[structopt(short = "-I", long, parse(from_os_str))]
+    include: Vec<PathBuf>,
+
+    /// If stubs for GRPC services should be generated or not (not implemented)
+    #[structopt(long)]
+    grpc: bool,
+
+    /// Fields that should be referenced from other crates, using prost-build's extern_path
+    /// functionality.  Each path should provide a "crate_name,file_descriptor_path" pair.  These
+    /// are separated by a comma.  The first part is the name of the crate that contains the
+    /// prost protobuf code, and the second part is the path to a serialized file descriptor set.
+    #[structopt(long, parse(try_from_str))]
+    external: Vec<ExternPathSetArg>,
+}
+
+/// A multi-line header to mark autogeneration of the file.  Includes a @generated tag for
+/// tools that support it.
+const AUTOGENERATION_HEADER: &str = "// DO NOT EDIT: Autogenerated by prostgen\n\
+                                     // @generated\n\n";
+
+impl ProstGen {
+    /// Execute the proto generation logic per the passed in command-line arguments.
+    /// This will result in a single |output| file being generated from the |input| files specified.
+    /// Consumes itself as a way of being conservative for any future changes that might be
+    /// non-deterministic.  This could reasonably borrow self without consequence.
+    pub fn run(self) -> anyhow::Result<()> {
+        let config = tonic_build::configure()
+            .build_client(self.grpc)
+            .build_server(self.grpc)
+            .format(true)
```

Comment:
I’ve only skimmed the build, but I didn’t see a data dep anywhere; is
this hermetic? (If so, I’d love to steal your incantation!)


cf. <https://github.com/bazelbuild/rules_rust/issues/388#issuecomment-725837114>

Reply:
Ah! I see now that the `_prost_generator` rule is properly wired up to
the toolchain. I was hoping that there was a way that didn’t require
having to do that :-) but that should definitely work.